examples/random_walk.py

```diff
@@ -77,22 +65,25 @@ def run_random_walk(env: CompilerEnv, step_count: int) -> List[float]:
                 break
         env.close()
 
-    def reward_delta(reward):
-        delta = rewards[0] / max(reward, 1e-9) - 1
-        return emph(f"{'+' if delta >= 0 else ''}{delta:.2%}")
+    def reward_percentage(reward, rewards):
+        if sum(rewards) == 0:
+            return 0
+        percentage = reward / sum(rewards)
+        return emph(f"{'+' if percentage >= 0 else ''}{percentage:.2%}")
 
     print(
         f"\nCompleted {emph(humanize.intcomma(step_num))} steps in {episode_time} "
         f"({step_num / episode_time.time:.1f} steps / sec)."
     )
-    print(f"Init reward:  {rewards[0]}")
-    print(f"Final reward: {rewards[-1]} ({reward_delta(rewards[-1])})")
+    print(f"Total reward: {sum(rewards)}")
     print(
-        f"Max reward:   {max(rewards)} ({reward_delta(max(rewards))} "
+        f"Max reward:   {max(rewards)} ({reward_percentage(max(rewards), rewards)} "
         f"at step {humanize.intcomma(rewards.index(max(rewards)) + 1)})"
     )
-    return rewards
-
+    def remove_no_change(rewards, actions):
+        return [a for (r, a) in zip(rewards, actions) if r != 0]
```

Comment:
I like the idea of reporting "effective actions", but we should be careful about how "effective" is defined. It would be quite possible for an action to have an effect on the environment that does not yield a reward, but changes the state in a way that meaningfully affects the environment. E.g. if the reward space is instruction count, consider an action that both removes 10 instructions and then adds 10 instructions. It has certainly affected the environment, but will produce a reward of zero.

If the goal of `remove_no_change()` is to filter the list of actions but still produce the same final state, that will be a little bit more tricky, but would be extremely useful.

One thing that might be helpful for this is the `action_had_no_effect` element that is returned in the `info` dict of all LLVM environments:

```py
>>> observation, reward, done, info = env.step(action)
>>> info
{'action_had_no_effect': True, 'new_action_space': False}
```

Although worth noting that this value is produced by LLVM and it actually under-reports changes, i.e. it is possible for an action to modify the state but for LLVM to report that it did nothing. For the LLVM environments, the most-complete view of the state can be obtained by hashing the LLVM-IR of the environment:

```py
>>> import hashlib
>>> hashlib.sha1(env.ir.encode("utf-8")).hexdigest()
'24c6dd4a6d4cb8dfb818577070432f41d4cd78b1'
```

Tracking how that hash changes over the course of a trajectory may well provide a more accurate view on which actions had an effect than the reward values.

Consider also that there are potentially subsequences of actions that effectively cancel each other out, so those too could be pruned. For example, the LLVM actions `-mem2reg` and `-reg2mem` have opposite effects:

```
$ python -m compiler_gym.bin.manual_env --env=llvm-ic-v0 --benchmark=cBench-v0/dijkstra
Initialized environment in 111.6ms
Welcome to the CompilerGym Shell!
---------------------------------
Type help or ? for more information.
The 'tutorial' command will give a step by step guide.

compilergym:cBench-v0/dijkstra> reset
Reset in 85.2ms

compilergym:cBench-v0/dijkstra> action -mem2reg
Action -mem2reg
Reward: 0.625806
Actions -mem2reg in 1.1ms with reward 0.6258064516129033.

compilergym:cBench-v0/dijkstra> action -reg2mem
Action -reg2mem
Reward: -0.703226
Actions -reg2mem in 909.6us with reward -0.7032258064516129.

compilergym:cBench-v0/dijkstra> action -mem2reg
Action -mem2reg
Reward: 0.625806
Actions -mem2reg in 1.1ms with reward 0.6258064516129033.

...

compilergym:cBench-v0/dijkstra> stack
   Depth | Action   | Effect   | Done   |    Reward |   Cumulative Reward
---------+----------+----------+--------+-----------+---------------------
       8 | -reg2mem | True     | False  | -0.664516 |           -0.193548
       7 | -mem2reg | True     | False  |  0.625806 |            0.470968
       6 | -reg2mem | True     | False  | -0.664516 |           -0.154839
       5 | -mem2reg | True     | False  |  0.625806 |            0.509677
       4 | -reg2mem | True     | False  | -0.664516 |           -0.116129
       3 | -mem2reg | True     | False  |  0.625806 |            0.548387
       2 | -reg2mem | True     | False  | -0.703226 |           -0.077419
       1 | -mem2reg | True     | False  |  0.625806 |            0.625806
       0 | <init>   | False    | False  |  0        |            0
```

It could be that `remove_no_change()` could detect and remove these redundant subsequences.

Of course this is well beyond the scope of this PR so feel free to ignore me :)

Reply:
Aye,

I'd like there to be `strict` mode for `simplify-stack` in the shell. If you end up building something like that, can you let me know?
If there is no reward set, it should pull the IRs and do the hashing like Chris says.  I'm not 100% sure that even that is enough - does LLVM keep anything sneaky anywhere? Still can't do anything about it if it does.
If there is a reward set, it should remove actions while that doing so increases or does not change the final reward. The current 'non strict' version just removes any passes that didn't have a positive reward last time. This isn't the same thing though, since a negative pass may be needed to set up a positive pass later. I was thinking a greedy approach would be sufficient, with multiple runs possible since one round of simplification changes the stack and may allow another run to have more effect (This definitely happens already with the non strict version).

[totally unimportant point] Chris, btw, 
`action -mem2reg -reg2mem -mem2reg -reg2mem -mem2reg -reg2mem -mem2reg -reg2mem`
Does the same thing as the 8 calls to action.